src/contracts/core/AVSDirectory.sol

```diff
@@ -52,137 +61,662 @@ contract AVSDirectory is
+     * @param completableTimestamp timestamp when the queued deallocation can be completed and freeMagnitude updated
+     */
+    function _queueDeallocate(
+        address operator,
+        MagnitudeAdjustment calldata deallocation,
+        uint32 completableTimestamp
+    ) internal {
+        IStrategy strategy = deallocation.strategy;
+        require(
+            deallocation.operatorSets.length == deallocation.magnitudeDiffs.length,
+            "AVSDirectory.queueDeallocation: operatorSets and magnitudeDiffs length mismatch"
+        );
+        OperatorSet[] calldata operatorSets = deallocation.operatorSets;
+
+        // deallocate from each operator set for this strategy
+        for (uint256 i = 0; i < operatorSets.length; ++i) {
+            // 1. ensure deallocation doesn't exceed current allocation
+            uint64 currentMagnitude = uint64(
+                _magnitudeUpdate[operator][strategy][operatorSets[i].avs][operatorSets[i].operatorSetId]
+                    .upperLookupRecent(uint32(block.timestamp))
+            );
+            require(
+                deallocation.magnitudeDiffs[i] <= currentMagnitude,
+                "AVSDirectory.queueDeallocation: cannot deallocate more than what is allocated"
+            );
+
+            // 2. update and decrement current and future queued amounts in case any pending allocations exist
+            _magnitudeUpdate[operator][strategy][operatorSets[i].avs][operatorSets[i].operatorSetId]
+                .decrementCurrentAndFutureCheckpoints({
+                timestamp: uint32(block.timestamp),
+                decrementValue: deallocation.magnitudeDiffs[i]
+            });
+
+            // TODO: ensure only 1 queued deallocation per operator, operatorSet, strategy
+            // _verifySinglePendingDeallocation();
+
+            // 3. queue deallocation for (op, opSet, strategy) for magnitudeDiff amount
+            _queuedDeallocations[operator][strategy][operatorSets[i].avs][operatorSets[i].operatorSetId].push(
+                QueuedDeallocation({
+                    magnitudeDiff: deallocation.magnitudeDiffs[i],
+                    completableTimestamp: completableTimestamp
+                })
+            );
+        }
+    }
+
+    /**
+     * @notice Completes queued deallocations for a single strategy and multiple operator sets
+     * @param operator address to complete deallocations for
+     * @param strategy the strategy to complete deallocations for
+     * @param operatorSet the operator set to complete deallocations for the given strategy
+     */
+    function _completeDeallocation(
+        address operator,
+        IStrategy strategy,
+        OperatorSet calldata operatorSet
+    ) internal returns (uint64 freeMagnitudeToAdd) {
+        QueuedDeallocation[] memory queuedDeallocation =
+            _queuedDeallocations[operator][strategy][operatorSet.avs][operatorSet.operatorSetId];
```

Comment:
how does this work if you set this to memory?

Reply:
we simply read the freeMagnitude to add back and add to running sum. We just update the `_nextDeallocationIndex` for the next time completeDeallocations is called